src/io/recv_multi.rs

```diff
@@ -0,0 +1,87 @@
+use crate::buf::bufgroup::BufX;
+use crate::io::SharedFd;
+
+use crate::buf::bufring::BufRing;
+use crate::runtime::driver::op::{CqeResult, MultiCQEStream, Op, Streamable};
+use crate::runtime::CONTEXT;
+use std::io;
+
+pub(crate) type RecvMultiStream = Op<RecvMulti, MultiCQEStream>;
+
+pub(crate) struct RecvMulti {
```

Comment:
Have you thought about the generalized verions `recvmsg_multi` too? I'm curious how the API works when receiving a single packet into multiple buffers

Reply:
I don't follow the question. You've seen Luca's excellent contribution to io-uring? 
- https://github.com/tokio-rs/io-uring/pull/192

Each recvmsg result is going into a single buffer, pulled from the buf_ring pool by the kernel. What am I missing?

Reply:
Does it unconditional _have_ too? [recvmsg](https://linux.die.net/man/2/recvmsg) takes an array of iovec. See PR #251 for the non multi-shot version, and also the existing sendmsg implementation. I read the docs around `io_uring_recvmsg_out` as an extra on top of the provided iovec specification

Reply:
I really don't know where you are coming from with this. Can you refer to the liburing documentation to point out what you think I'm missing? The 251 PR doesn't seem relevant to how kernel io_uring implements multishot recvmsg at all?

Reply:
> I really don't know where you are coming from with this. Can you refer to the liburing documentation to point out what you think I'm missing? The 251 PR doesn't seem relevant to how kernel io_uring implements multishot recvmsg at all?

My apologies: #252 is relevant. #251 is completely irrelevant

Reply:
If you linked to something by mistake, can you edit your comments to remove the link, at least the pound sign?

Reply:
> I updated the original link to point to the correct PR

Reply:
And my point is the multi recvmsg doesn't do multiple buffers - it  couldn't. Luca made that clear with his investigation and his comments.

Reply:
> I updated the original link to point to the correct PR

251 is still preceded by a hash in one of your comments. Now I'm totally wondering why all these PRs have to be linked with each other.

Reply:
> And my point is the multi recvmsg doesn't do multiple buffers - it couldn't. Luca made that clear with his investigation and his comments.

That may well be true, I'm really not an expert here, but I'm not able to see that spelt out definitely in that PR or the docs.

I have tried tracing the kernel source, but I'm none the wiser either way on where that limitation comes into force.

Reply:
> The question was, have you considered how to cope with streaming operations where multiple buffers are provided in the original msghdr: that implies a collections of buffers

I have considered that collections of buffers are not part of the io_uring multishot receiving APIs so we don't need a streaming API for such an animal.

Reply:
> I have considered that collections of buffers are not part of the io_uring multishot receiving APIs so we don't need a streaming API for such an animal.

Thats a shame: that's a difference in API vs the non multishot API and the linux syscall, and precludes some of the memory locality use cases i spelt out in #252

Reply:
I don't wish to detract from this pr: I like this a great deal, especially the provided ring buffers required to make it work.

I'm looking at this API, as with most, with my use case in mind, which is UDP packet processing on a woefully underpowered aarch64 platform. Outperforming recv_mmsg is what brought me to this crate in the first place. 

Personally I'm not invested in tokio api compliance at all - I simply don't care if we completely diverge. The API I always compare to are the base syscalls